README.md
<!-- badges: start -->
[![Build Status](https://travis-ci.com/Cristianetaniguti/onemap.svg?branch=master)](https://travis-ci.com/Cristianetaniguti/onemap) 
[![R-CMD-check](https://github.com/Cristianetaniguti/onemap/workflows/R-CMD-check/badge.svg)](https://github.com/Cristianetaniguti/onemap/actions)
[![Development](https://img.shields.io/badge/development-active-blue.svg)](https://img.shields.io/badge/development-active-blue.svg)
[![codecov](https://codecov.io/gh/Cristianetaniguti/onemap/branch/master/graph/badge.svg)](https://codecov.io/gh/Cristianetaniguti/onemap)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/onemap)](https://cran.r-project.org/package=onemap)
[![CRAN_monthly_downloads](https://cranlogs.r-pkg.org/badges/onemap)](https://cranlogs.r-pkg.org/badges/onemap)
  <!-- badges: end -->
  
# OneMap <img src="https://user-images.githubusercontent.com/7572527/119237022-0b19a400-bb11-11eb-9d45-228a59f22a1a.png" align="right" width="200"/>

`OneMap` is a software for constructing genetic maps in experimental
crosses: full-sib, RILs, F2, and backcrosses. It was developed by
Gabriel R A Margarido, Marcelo Mollinari and A Augusto F Garcia. Later on, 
Rodrigo R Amadeu, Cristiane H Taniguti, and Getulio C. Ferreira joined the project.

It has been available on CRAN for several years
(https://cran.r-project.org/package=onemap). Its last version was
updated on 2020-02-17. CRAN has OneMap's stable version, which is
recommended for most users.

This GitHub page has its version under development. New functions will
be added (experimental work) and, once it is done, we will synchronize
the repositories and add them to CRAN.

We worked very hard to release a new stable version allowing users to
analyze data sets with markers based on sequencing technologies, such
as Illumina, GBS, etc.

`OneMap` comprises a set of functions that allow users to build a
linkage map. Some functions are used internally by the package, and
should not be used directly.

# How to install

## From CRAN (stable version)

It is easy, just type (within R):

```R
setRepositories(ind = 1:2)
install.packages("onemap", dependencies=TRUE)
```

You also can use the console menus: _Packages -> Install
package(s)_. After clicking, a box will pop-up asking you to choose
the CRAN mirror. Choose the location nearest to you. Then, another box
will pop-up asking you to choose the package you want to install.
Select _onemap_ then click _OK_. The package will be
automatically installed on your computer.

`OneMap` can also be installed by downloading the appropriate files
directly at the CRAN website and following the instructions given in section `6.3 Installing Packages` of the
[R Installation and Administration](https://cran.r-project.org/doc/manuals/R-admin.pdf)
manual.

## From github (version under development)

Within R, you need to install and load the package `devtools`:

```R
install.packages("devtools")
library(devtools)
```

This will allow you to automatically build and install packages from
GitHub. If you use Windows, first install
[Rtools](https://cran.r-project.org/bin/windows/Rtools/). If you are facing problems with Rtools installation, try to do it by selecting the *Run as Administrator* option with the right mouse button. On a Mac,
you will need Xcode (available on the App Store).

Then, to install `OneMap` from GitHub (this very repo):

```R
install_github("augusto-garcia/onemap")
```

## From docker hub

`OneMap` requires several dependencies that you may not have in your system. To overcome the need of installing all of them, you can use the `OneMap` image in the docker hub. Install docker (see more about [here](https://docs.docker.com/get-started/)) and use:

```bash
docker pull cristaniguti/onemap_git:latest
```

The `OneMap` image already has the RStudio from rocker image, you can run it in your favorite browser running the following command:

```bash
docker run -p 8787:8787 -v $(pwd):/home/rstudio/ -e DISABLE_AUTH=true cristaniguti/onemap_git
```

This will make the container available in port 8787 (choose other if you prefer). The `-v` argument includes directories of your computer, in this case, the current directory (pwd) to the container. You can use `-v` several times to include several directories. After, you just need to go to your favorite browser and search for <your_localhost>:8787 (example 127.0.0.1:8787). That is it! Everything you need is there.

# Tutorials

You can read _OneMap_ tutorials going to the vignettes of the
installed package, or clicking below. Please, start with the overview,
that will guide you through other chapters.

1. [Overview](https://statgen-esalq.github.io/tutorials/onemap/Overview.html)

2. [Introduction to R](https://statgen-esalq.github.io/tutorials/onemap/Introduction_R.html)

3. [How to build a linkage map for inbred-bases populations (F2, RIL and BC)](https://statgen-esalq.github.io/tutorials/onemap/Inbred_Based_Populations.html)

4. [How to build a linkage map for outcrossing populations](https://statgen-esalq.github.io/tutorials/onemap/Outcrossing_Populations.html)

5. [A guide to build high-density linkage maps](https://cristianetaniguti.github.io/Tutorials/onemap/Quick_HighDens/High_density_maps.html)

# How to cite

Margarido, G. R. A., Souza, A. P., &38; Garcia, A. A. F. (2007). OneMap: software for genetic mapping in outcrossing species. Hereditas, 144(3), 78–79. https://doi.org/10.1111/j.2007.0018-0661.02000.x

* If you are using OneMap versions > 2.0, please cite also:

Taniguti, C. H.; Taniguti, L. M.; Amadeu, R. R.; Lau, J.; de Siqueira Gesteira, G.; Oliveira, T. de P.; Ferreira, G. C.; Pereira, G. da S.;  Byrne, D.;  Mollinari, M.; Riera-Lizarazu, O.; Garcia, A. A. F. Developing best practices for genotyping-by-sequencing analysis in the construction of linkage maps. GigaScience, 12, giad092. https://doi.org/10.1093/gigascience/giad092

* If you used the HMM parallelization, please cite [BatchMap](https://github.com/bschiffthaler/BatchMap) paper too:

Schiffthaler, B., Bernhardsson, C., Ingvarsson, P. K., &; Street, N. R. (2017). BatchMap: A parallel implementation of the OneMap R package for fast computation of F1 linkage maps in outcrossing species. PLoS ONE, 12(12), 1–12. https://doi.org/10.1371/journal.pone.0189256

NEWS.md
# onemap 2.0.5

* Organizing the files in directories to allow installation from github 
* Moving functions descriptions for roxygen style
* Solving conflict of version 2.0.4 with 2.0.5 (the one in CRAN)
* Replacing old vignettes in pdf with new ones using RMarkdown
* Adding functions to plot raw data
* Adding functions to draw graphics showing marker categories
* Adding functions to perform chi-square tests for marker segregation
* Adding functions to identify alpha under Bonferroni's argument
* Correcting documentation not compiling for RMarkdown
* Updating dependencies for other R packages

# onemap 2.0.6

* (To be included)

# onemap 2.0.7

* Updating .travis.yml for the newer standards
* Updating README in order to solve Windows user's issues.
* Updating dependencies and NAMESPACE for newer standards (R-3.4.0)
* Updating dependencies within Bioconductor packages

# onemap 2.0.8

* Fixing chi-square segregation tests for markers with missing classes

# onemap 2.1.1001

* Bug fix in function test_segregation for F2 cross
* Fix in example f2
* Codif_data changed to deal with dominant markers for F2
* Warnings removed and labels added to outcross plot
* plot.onemap palette changed
* Included C.A, D.B and A.H.B types in plot by segregation
* New argument in select_segreg
* Function create_data_bins now keeps CHROM and POS infos
* Possible to make sequences from CHROM and POS using 2pts object
* Outcrossing vcf example file are available
* New functions group_seq and print.group_seq
* Functions draw.try and draw_order are now defuncts
* Vignette for outcrossing updated
* F2 vcf example file are available
* Vignette for inbred updated

# onemap 2.1.1002

* onemap_read_vcfR function
* Examples files for rils and backcross populations 
* Vignette for inbred and outcrossing updated

# onemap 2.1.1003

* rf_graph_table updated with ggplot2 and plotly
* NAMESPACE automated with roxygen2

# onemap 2.1.1004

* write_onemap_raw function

# onemap 2.1.1005

* Vignette updated

# onemap 2.1.1006

* vcf2raw as defunct

# onemap 2.1.1007

* draw_map2 function
* Vignette update

# onemap 2.1.1008

* mds_onemap function
* Vignette update

# onemap.2.1.2

* Submitted to CRAN

# onemap 2.1.2001

* Adapt compare to rils and backcross

# onemap 2.1.3

* draw_map2 now receives data.frames
* Changing class functions to is
* Adapt try_seq to rils and backcross
* onemap_read_vcfR also accept phased VCF
* Submitted to CRAN

# onemap 2.2.0

* split_onemap function to split onemap objects
* filter_missing function to filter markers in onemap object by missing data
* map parallelized with Batchmap method
* Option rm_unlinked in map to remove markers which not reach the linkage criterias when estimating genetic distances by HMM
* All ordering algorithms now presents the possibility to parallelize the last step (map) and remove automatically unlinked markers and repeat the ordering procedure
* suggest_lod function now works also with sequence objects
* Remove argument hmm from mds_onemap
* Solve numerical problems with small negative LOD values in rf_2pts 
* Update all RDatas to new format and compressed form
* Function map_avoid_unlinked to automatically remove the unlinked markers and restart the map
* create_depths_profile plots allele counts and genotypes from vcf, onemap and errors
* runpedsim  wrapper function for PedigreeSim software
* pedsim2raw converts PedigreeSim outputs to onemap raw file
* pedsim2vcf converts PedigreeSim outputs to VCF file simulating counts with updog or negative binomial
* Replace all get() by nothing
* Export parents and progeny haplotypes with write_haplotypes.R
* Draw graphics for progeny haplotypes
* F2 intercross populations now uses the same HMM than outcrossing populations. Modification needed to correct infer dominant markers and also to export progeny haplotypes.
* Function vcf2progeny_haplotypes to convert phased VCF in onemap_progeny_haplotypes object. It make possible to draw the haplotypes for phased VCFs.
* Function seq_by_type split sequence by marker type
* Internal function map_save_ram creates new onemap object for only the evaluated sequence and, after finished process that require too much RAM memory, it return the complete onemap object to user. This save RAM memory in parallelization process.
* Vignettes update
* Documentation update

# onemap 2.3.0

* Remove functions to perform simulations

R/compare.R
#######################################################################
##                                                                     ##
## Package: onemap                                                     ##
##                                                                     ##
## File: compare.R                                                     ##
## Contains: compare, print.compare                                    ##
##                                                                     ##
## Written by Gabriel R A Margarido & Marcelo Mollinari                ##
## copyright (c) 2009, Gabriel R A Margarido & Marcelo Mollinari       ##
##                                                                     ##
## First version: 02/27/2009                                           ##
## License: GNU General Public License version 2 (June, 1991) or later ##
##                                                                     #
#######################################################################

##' Compare all possible orders (exhaustive search) for a given sequence of
##' markers
##'
##' For a given sequence with \eqn{n}{n} markers, computes the multipoint
##' likelihood of all \eqn{\frac{n!}{2}}{n!/2} possible orders.
##'
##' Since the number \eqn{\frac{n!}{2}}{n!/2} is large even for moderate values
##' of \eqn{n}{n}, this function is to be used only for sequences with
##' relatively few markers. If markers were genotyped in an outcross population,
##' linkage phases need to be estimated and therefore more states need to be
##' visited in the Markov chain; when segregation types are D1, D2 and C,
##' computation can required a very long time (specially when markers linked in
##' repulsion are involved), so we recommend to use this function up to 6 or 7 markers.
##' For inbred-based populations, up to 10 or 11 markers can be ordered with this function,
##' since linkage phase are known.
##' The multipoint likelihood is calculated according to Wu et al.
##' (2002b) (Eqs. 7a to 11), assuming that the recombination fraction is the
##' same in both parents. Hidden Markov chain codes adapted from Broman et al.
##' (2008) were used.
##' 
##' @importFrom methods is
##' @importFrom utils setTxtProgressBar txtProgressBar tail head flush.console
##'
##' @param input.seq an object of class \code{sequence}.
##' @param n.best the number of best orders to store in object (defaults to
##' 50).
##' @param tol tolerance for the C routine, i.e., the value used to evaluate
##' convergence.
##' @param verbose if \code{FALSE} (default), simplified output is displayed.
##' if \code{TRUE}, detailed output is displayed.
##' @return An object of class \code{compare}, which is a list containing the
##' following components: \item{best.ord}{a \code{matrix} containing the best
##' orders.} \item{best.ord.rf}{a \code{matrix} with recombination frequencies
##' for the corresponding best orders.} \item{best.ord.phase}{a \code{matrix}
##' with linkage phases for the best orders.} \item{best.ord.like}{a
##' \code{vector} with log-likelihood values for the best orders.}
##' \item{best.ord.LOD}{a \code{vector} with LOD Score values for the best
##' orders.} \item{data.name}{name of the object of class \code{onemap} with
##' the raw data.} \item{twopt}{name of the object of class \code{rf_2pts} with
##' the 2-point analyses.}
##' @author Marcelo Mollinari, \email{mmollina@@usp.br}
##' @seealso \code{\link[onemap]{marker_type}} for details about segregation
##' types and \code{\link[onemap]{make_seq}}.
##' @references Broman, K. W., Wu, H., Churchill, G., Sen, S., Yandell, B.
##' (2008) \emph{qtl: Tools for analyzing QTL experiments} R package version
##' 1.09-43
##'
##' Jiang, C. and Zeng, Z.-B. (1997). Mapping quantitative trait loci with
##' dominant and missing markers in various crosses from two inbred lines.
##' \emph{Genetica} 101: 47-58.
##'
##' Lander, E. S., Green, P., Abrahamson, J., Barlow, A., Daly, M. J., Lincoln,
##' S. E. and Newburg, L. (1987) MAPMAKER: An interactive computer package for
##' constructing primary genetic linkage maps of experimental and natural
##' populations. \emph{Genomics} 1: 174-181.
##'
##' Mollinari, M., Margarido, G. R. A., Vencovsky, R. and Garcia, A.  A. F.
##' (2009) Evaluation of algorithms used to order markers on genetics maps.
##' _Heredity_ 103: 494-502.
##'
##' Wu, R., Ma, C.-X., Painter, I. and Zeng, Z.-B. (2002a) Simultaneous maximum
##' likelihood estimation of linkage and linkage phases in outcrossing species.
##' \emph{Theoretical Population Biology} 61: 349-363.
##'
##' Wu, R., Ma, C.-X., Wu, S. S. and Zeng, Z.-B. (2002b). Linkage mapping of
##' sex-specific differences. \emph{Genetical Research} 79: 85-96
##' @keywords utilities
##' @examples
##'
##' \donttest{
##'   #outcrossing example
##'   data(onemap_example_out)
##'   twopt <- rf_2pts(onemap_example_out)
##'   markers <- make_seq(twopt,c(12,14,15,26,28))
##'   (markers.comp <- compare(markers))
##'   (markers.comp <- compare(markers,verbose=TRUE))
##'
##'   #F2 example
##'   data(onemap_example_f2)
##'   twopt <- rf_2pts(onemap_example_f2)
##'   markers <- make_seq(twopt,c(17,26,29,30,44,46,55))
##'   (markers.comp <- compare(markers))
##'   (markers.comp <- compare(markers,verbose=TRUE))
##'
##' }
##'
##'@export

compare<- function(input.seq,n.best=50,tol=10E-4,verbose=FALSE) {
  if(inherits(input.seq$data.name, c("outcross", "f2")))
    return(compare_outcross(input.seq=input.seq,n.best=n.best,tol=tol,verbose=verbose))
  else 
    return(compare_inbred_bc(input.seq=input.seq,n.best=n.best,tol=tol,verbose=verbose))
}

## Compare all possible orders (exhaustive search) for a given sequence of
## markers (for outcrosses)
compare_outcross<- function(input.seq, n.best=50, tol=10E-4, verbose=FALSE)
{
  ## checking for correct objects
  if(!inherits(input.seq,"sequence"))
    stop(sQuote(deparse(substitute(input.seq)))," is not an object of class 'sequence'")
  if(length(input.seq$seq.num) > 5)
    warning("This operation may take a VERY long time\n")
  flush.console()
  if(length(input.seq$seq.num) > 10) {
    warning("It is not wise trying to use 'compare' with more than 10 markers")
    ANSWER <- readline("Are you sure you want to proceed? [y or n]\n")
    while(substr(ANSWER, 1, 1) != "n" & substr(ANSWER, 1, 1) != "y")
      ANSWER <- readline("\nPlease answer: 'y' or 'n' \n")
    if (substr(ANSWER, 1, 1) == "n") stop("Execution stopped!")
  }
  if(length(input.seq$seq.num) == 2)
    return(map(input.seq, tol=tol)) ## nothing to be done for 2 markers
  else {
    ## allocating variables
    rf.init <- vector("list",length(input.seq$seq.num)-1)
    phase.init <- vector("list",length(input.seq$seq.num)-1)
    best.ord <- matrix(NA,(n.best+1),length(input.seq$seq.num))
    best.ord.rf <- matrix(NA,(n.best+1),length(input.seq$seq.num)-1)
    best.ord.phase <- matrix(NA,(n.best+1),length(input.seq$seq.num)-1)
    best.ord.like <- best.ord.LOD <- rep(-Inf,(n.best+1))
    
    ## 'phases' gathers information from two-point analyses
    list.init <- phases(input.seq)
    
    ## 'perm_pars' generates all n!/2 orders
    all.ord <- perm_pars(input.seq$seq.num)
    if (verbose){
      cat("\nComparing",nrow(all.ord),"orders:     \n\n")
      for(i in 1:nrow(all.ord)){
        ## print output for each order
        cat("Order", i, ":", all.ord[i,], "\n")
        flush.console()
        ## get initial values for the HMM
        all.match <- match(all.ord[i,],input.seq$seq.num)
        for(j in 1:(length(input.seq$seq.num)-1)){
          if(all.match[j] > all.match[j+1]){
            rf.init[[j]] <- list.init$rf.init[[acum(all.match[j]-2)+all.match[j+1]]]
            phase.init[[j]] <- list.init$phase.init[[acum(all.match[j]-2)+all.match[j+1]]]
          }
          else {
            rf.init[[j]] <- list.init$rf.init[[acum(all.match[j+1]-2)+all.match[j]]]
            phase.init[[j]] <- list.init$phase.init[[acum(all.match[j+1]-2)+all.match[j]]]
          }
        }
        Ph.Init <- comb_ger(phase.init)
        Rf.Init <- comb_ger(rf.init)
        if(nrow(Ph.Init)>1){
          ##Removing ambigous phases
          rm.ab<-rem_amb_ph(M=Ph.Init, w=input.seq, seq.num=all.ord[i,])
          Ph.Init <- Ph.Init[rm.ab,]
          Rf.Init <- Rf.Init[rm.ab,]
          if(inherits(Ph.Init,"integer")){
            Ph.Init<-matrix(Ph.Init,nrow=1)
            Rf.Init<-matrix(Rf.Init,nrow=1)
          }
        }
        for(j in 1:nrow(Ph.Init)){
          ## estimate parameters
          final.map <- est_map_hmm_out(geno=t(input.seq$data.name$geno[,all.ord[i,]]),
                                       error=input.seq$data.name$error[all.ord[i,] + rep(c(0:(input.seq$data.name$n.ind-1))*input.seq$data.name$n.mar, each=length(all.ord[i,])),],
                                       type=input.seq$data.name$segr.type.num[all.ord[i,]],
                                       phase=Ph.Init[j,],
                                       rf.vec=Rf.Init[j,],
                                       verbose=FALSE,
                                       tol=tol)
          best.ord[(n.best+1),] <- all.ord[i,]
          best.ord.rf[(n.best+1),] <- final.map$rf
          best.ord.phase[(n.best+1),] <- Ph.Init[j,]
          best.ord.like[(n.best+1)] <- final.map$loglike
          
          ## arrange orders according to the likelihood
          like.order <- order(best.ord.like, decreasing=TRUE)
          best.ord <- best.ord[like.order,]
          best.ord.rf <- best.ord.rf[like.order,]
          best.ord.phase <- best.ord.phase[like.order,]
          best.ord.like <- sort(best.ord.like, decreasing=TRUE)
        }
      }
    } else{
      count <- 0
      pb <- txtProgressBar(style=3)
      setTxtProgressBar(pb, 0)
      
      for(i in 1:nrow(all.ord)){
        all.match <- match(all.ord[i,],input.seq$seq.num)
        for(j in 1:(length(input.seq$seq.num)-1)){
          if(all.match[j] > all.match[j+1]){
            rf.init[[j]] <- list.init$rf.init[[acum(all.match[j]-2)+all.match[j+1]]]
            phase.init[[j]] <- list.init$phase.init[[acum(all.match[j]-2)+all.match[j+1]]]
          }
          else {
            rf.init[[j]] <- list.init$rf.init[[acum(all.match[j+1]-2)+all.match[j]]]
            phase.init[[j]] <- list.init$phase.init[[acum(all.match[j+1]-2)+all.match[j]]]
          }
        }
        Ph.Init <- comb_ger(phase.init)
        Rf.Init <- comb_ger(rf.init)
        if(nrow(Ph.Init)>1){
          ##Removing ambigous phases
          rm.ab<-rem_amb_ph(M=Ph.Init, w=input.seq, seq.num=all.ord[i,])
          Ph.Init <- Ph.Init[rm.ab,]
          Rf.Init <- Rf.Init[rm.ab,]
          if(inherits(Ph.Init,"integer")){
            Ph.Init<-matrix(Ph.Init,nrow=1)
            Rf.Init<-matrix(Rf.Init,nrow=1)
          }
        }
        for(j in 1:nrow(Ph.Init)){
          ## estimate parameters
          final.map <- est_map_hmm_out(geno=t(input.seq$data.name$geno[,all.ord[i,]]),
                                       error=input.seq$data.name$error[all.ord[i,] + rep(c(0:(input.seq$data.name$n.ind-1))*input.seq$data.name$n.mar, each=length(all.ord[i,])),],
                                       type=input.seq$data.name$segr.type.num[all.ord[i,]],
                                       phase=Ph.Init[j,],
                                       rf.vec=Rf.Init[j,],
                                       verbose=FALSE,
                                       tol=tol)
          best.ord[(n.best+1),] <- all.ord[i,]
          best.ord.rf[(n.best+1),] <- final.map$rf
          best.ord.phase[(n.best+1),] <- Ph.Init[j,]
          best.ord.like[(n.best+1)] <- final.map$loglike
          
          ## arrange orders according to the likelihood
          like.order <- order(best.ord.like, decreasing=TRUE)
          best.ord <- best.ord[like.order,]
          best.ord.rf <- best.ord.rf[like.order,]
          best.ord.phase <- best.ord.phase[like.order,]
          best.ord.like <- sort(best.ord.like, decreasing=TRUE)
        }
        count<-count+1
        setTxtProgressBar(pb, count/nrow(all.ord))
      }
      close(pb)
    }

    best.ord.LOD <- round((best.ord.like-max(best.ord.like))/log(10),4)
    structure(list(best.ord = best.ord,
                   best.ord.rf = best.ord.rf,
                   best.ord.phase = best.ord.phase,
                   best.ord.like = best.ord.like,
                   best.ord.LOD = best.ord.LOD,
                   data.name=input.seq$data.name,
                   twopt=input.seq$twopt), class = "compare")
    
  }
}

## Compare all possible orders (exhaustive search) for a given sequence of
## markers (crosses derived from inbred lines)
compare_inbred_bc<- function(input.seq, n.best=50, tol=10E-4, verbose=FALSE) 
{
  ## checking for correct objects
  if(!inherits(input.seq,"sequence"))
    stop(sQuote(deparse(substitute(input.seq)))," is not an object of class 'sequence'")
  if(length(input.seq$seq.num) > 5)
    warning("This operation may take a VERY long time\n")
  flush.console()
  if(length(input.seq$seq.num) > 10) {
    warning("It is not wise trying to use 'compare' with more than 10 markers")
    ANSWER <- readline("Are you sure you want to proceed? [y or n]\n")
    while(substr(ANSWER, 1, 1) != "n" & substr(ANSWER, 1, 1) != "y")
      ANSWER <- readline("\nPlease answer: 'y' or 'n' \n")
    if (substr(ANSWER, 1, 1) == "n") stop("Execution stopped!")
  }
  if(length(input.seq$seq.num) == 2)
    return(map(input.seq, tol=tol)) ## nothing to be done for 2 markers
  else
  {
    ## allocating variables
    all.ord <- perm_pars(input.seq$seq.num)
    rf.init <- matrix(NA,nrow(all.ord),length(input.seq$seq.num)-1)
    best.ord <- matrix(NA,(n.best+1),length(input.seq$seq.num))
    best.ord.rf <- matrix(NA,(n.best+1),length(input.seq$seq.num)-1)
    best.ord.like <- best.ord.LOD <- rep(-Inf,(n.best+1))
    if(verbose) {
      cat("\nComparing",nrow(all.ord),"orders:     \n\n")
    } else {
      count <- 0
      pb <- txtProgressBar(style=3)
      setTxtProgressBar(pb, 0)
    }
    for(i in 1:nrow(all.ord))
    {
      ## print output for each order
      if (verbose) cat("Order", i, ":", all.ord[i,], "\n")
      flush.console()
      seq.temp<-make_seq(input.seq$twopt, arg=all.ord[i,])
      rf.temp<- get_vec_rf_in(seq.temp, acum=FALSE)
      final.map<- est_map_hmm_bc(geno=t(input.seq$data.name$geno[,all.ord[i,]]),
                                error=input.seq$data.name$error[all.ord[i,] + 
                                                                  rep(c(0:(input.seq$data.name$n.ind-1))*input.seq$data.name$n.mar, 
                                                                      each=length(all.ord[i,])),],
                                rf.vec=rf.temp,
                                verbose=FALSE,
                                tol=tol)
      if(inherits(input.seq$data.name, c("riself", "risib")))
        final.map$rf<-adjust_rf_ril(final.map$rf,
                                    type=class(input.seq$data.name)[2],
                                    expand = FALSE)
      best.ord[(n.best+1),] <- all.ord[i,]
      best.ord.rf[(n.best+1),] <- final.map$rf
      best.ord.like[(n.best+1)] <- final.map$loglike
      ## arrange orders according to the likelihood
      like.order <- order(best.ord.like, decreasing=TRUE)
      best.ord <- best.ord[like.order,]
      best.ord.rf <- best.ord.rf[like.order,]
      best.ord.like <- sort(best.ord.like, decreasing=TRUE)
      if(!verbose)
      {
        count<-count+1
        setTxtProgressBar(pb, count/nrow(all.ord))
      }
    }
    close(pb)
    best.ord.LOD <- round((best.ord.like-max(best.ord.like))/log(10),4)
    structure(list(best.ord = best.ord,
                   best.ord.rf = best.ord.rf,
                   best.ord.phase = matrix(1, nrow(best.ord.rf), ncol(best.ord.rf)),
                   best.ord.like = best.ord.like,
                   best.ord.LOD = best.ord.LOD,
                   data.name=input.seq$data.name,
                   twopt=input.seq$twopt), class = "compare")
  }
}

## Compare all possible orders (exhaustive search) for a given sequence of
## markers (crosses derived from inbred lines)
compare_inbred_f2<- function(input.seq, n.best=50, tol=10E-4, verbose=FALSE) 
{
  ## checking for correct objects
  if(!inherits(input.seq,"sequence"))
    stop(sQuote(deparse(substitute(input.seq)))," is not an object of class 'sequence'")
  if(length(input.seq$seq.num) > 5)
    warning("This operation may take a VERY long time\n")
  flush.console()
  if(length(input.seq$seq.num) > 10) {
    warning("It is not wise trying to use 'compare' with more than 10 markers")
    ANSWER <- readline("Are you sure you want to proceed? [y or n]\n")
    while(substr(ANSWER, 1, 1) != "n" & substr(ANSWER, 1, 1) != "y")
      ANSWER <- readline("\nPlease answer: 'y' or 'n' \n")
    if (substr(ANSWER, 1, 1) == "n") stop("Execution stopped!")
  }
  if(length(input.seq$seq.num) == 2)
    return(map(input.seq, tol=tol)) ## nothing to be done for 2 markers
  else
  {
    ## allocating variables
    all.ord <- perm_pars(input.seq$seq.num)
    rf.init <- matrix(NA,nrow(all.ord),length(input.seq$seq.num)-1)
    best.ord <- matrix(NA,(n.best+1),length(input.seq$seq.num))
    best.ord.rf <- matrix(NA,(n.best+1),length(input.seq$seq.num)-1)
    best.ord.like <- best.ord.LOD <- rep(-Inf,(n.best+1))
    if(!verbose){
      count <- 0
      pb <- txtProgressBar(style=3)
      setTxtProgressBar(pb, 0)
    }
    for(i in 1:nrow(all.ord))
    {
      ## print output for each order
      if (verbose) cat("Order", i, ":", all.ord[i,], "\n")
      flush.console()
      seq.temp<-make_seq(input.seq$twopt, arg=all.ord[i,])
      seq.temp$twopt<-input.seq$twopt
      rf.temp<-get_vec_rf_in(seq.temp, acum=FALSE)
      final.map<-est_map_hmm_f2(geno=t(input.seq$data.name$geno[,all.ord[i,]]),
                                rf.vec=rf.temp,
                                verbose=FALSE,
                                tol=tol)
      best.ord[(n.best+1),] <- all.ord[i,]
      best.ord.rf[(n.best+1),] <- final.map$rf
      best.ord.like[(n.best+1)] <- final.map$loglike
      ## arrange orders according to the likelihood
      like.order <- order(best.ord.like, decreasing=TRUE)
      best.ord <- best.ord[like.order,]
      best.ord.rf <- best.ord.rf[like.order,]
      best.ord.like <- sort(best.ord.like, decreasing=TRUE)
      if(!verbose)
      {
        count<-count+1
        setTxtProgressBar(pb, count/nrow(all.ord))
      }
    }
    close(pb)
    best.ord.LOD <- round((best.ord.like-max(best.ord.like))/log(10),4)
    structure(list(best.ord = best.ord,
                   best.ord.rf = best.ord.rf,
                   best.ord.phase = matrix(1, nrow(best.ord.rf), ncol(best.ord.rf)),
                   best.ord.like = best.ord.like,
                   best.ord.LOD = best.ord.LOD,
                   data.name=input.seq$data.name,
                   twopt=input.seq$twopt), class = "compare")
  }
}

##' print method for object class 'compare'
##' 
##' @param x object of class compare
##' @param ... currently ignored
##' 
##' @return compare object description
##' 
##'@export
##'@method print compare
print.compare <- function(x,...) {
  FLAG<-0
  if(!(inherits(x$data.name, c("outcross","f2")))) FLAG<-1
  phases.char <- c("CC","CR","RC","RR")
  n.ord <- max(which(head(x$best.ord.LOD,-1) != -Inf))
  unique.orders <- unique(x$best.ord[1:n.ord,])
  n.ord.nest <- nrow(unique.orders)
  phases.nested <- vector("list",n.ord.nest)
  LOD <- vector("list",n.ord.nest)
  if(!FLAG){
    for (i in 1:n.ord.nest) {
      same.order <- which(apply(x$best.ord[1:n.ord,],1,function(x) all(x==unique.orders[i,])))
      ifelse(length(same.order)==1,phases.nested[[i]] <- t(as.matrix(x$best.ord.phase[same.order,])),phases.nested[[i]] <- x$best.ord.phase[same.order,])
      LOD[[i]] <- x$best.ord.LOD[same.order]
    }
  }
  skip <- c(nchar(n.ord.nest),max(nchar(unique.orders[1,])+2))
  cat("\nNumber of orders:",n.ord,"\n")
  if(FLAG==0){ ## outcrossing
    leng.print <- nchar(paste("order ",format(n.ord.nest,width=skip[1]),":  ",paste(format(unique.orders[1,],width=skip[2]),collapse=""),"     ",format(11.11,digits=2,format="f",width=6),"     ",format(11.11,digits=2,format="f",width=6),"\n",sep=""))
    cat(paste("Best ",n.ord.nest," unique orders",paste(rep(" ",leng.print-37),collapse=""),"LOD    Nested LOD","\n",sep=""))
    cat(paste(rep("-",leng.print),collapse=""),"\n")
  }
  else if(FLAG==1){ ## other
    leng.print <- nchar(paste("order ",format(n.ord.nest,width=skip[1]),":  ",paste(format(unique.orders[1,],width=skip[2]),collapse=""),"     ",format(11.11,digits=2,format="f",width=6),"\n",sep=""))
    cat(paste("Best ",n.ord.nest," unique orders",paste(rep(" ",leng.print-25),collapse=""),"LOD","\n",sep=""))
    cat(paste(rep("-",leng.print),collapse=""),"\n")
  }
  else stop ("Should not get here!")
  if(FLAG==0){ ## outcrossing
    for (i in 1:n.ord.nest) {
      cat(paste("order ",format(i,width=skip[1]),":  ",paste(format(unique.orders[i,],width=skip[2]),collapse=""),"\n",sep=""))
      for (j in 1:dim(phases.nested[[i]])[1]) {
        cat(paste("\t",paste(rep(" ",1+skip[1]+skip[2]),collapse=""),paste(format(phases.char[phases.nested[[i]][j,]],width=skip[2]),collapse=""),"     ",formatC(round(LOD[[i]][j],2),digits=2,format="f",width=6),"     ",formatC(round(LOD[[i]][j]-LOD[[i]][1],2),digits=2,format="f",width=6),"\n",sep=""))
      }
      cat(paste(rep("-",leng.print),collapse=""))
      cat("\n")
    }
  }
  else if(FLAG==1){ ## other
    for (i in 1:n.ord.nest) {
      cat(paste("order ",format(i,width=skip[1]),":  ",paste(format(unique.orders[i,],width=skip[2]),collapse=""), "     ",formatC(round(x$best.ord.LOD[i],2),digits=2,format="f",width=6),  "\n",sep=""))
    }
    cat(paste(rep("-",leng.print),collapse=""))
    cat("\n")
  }
  else stop ("Should not get here!")
}
## end of file
